File: 5-longest-palindromic-substring/5-longest-palindromic-substring.py
class Solution(object):
    def longestPalindrome(self, string):
        """
        :type string: str
        :rtype: str
        """
        ansIdx = [0, 1]
        for i in range(1, len(string)):
            odd = self.getLongestPalindrom(string, i-1, i+1)
            even = self.getLongestPalindrom(string, i-1, i)
            longestIdx = max(odd, even, key=lambda x: x[1] - x[0])
            ansIdx = max(longestIdx, ansIdx, key=lambda x: x[1] - x[0])
        return string[ansIdx[0] : ansIdx[1]]
    
    def getLongestPalindrom(self, string, left, right):
        while left >= 0 and right < len(string):
            if string[left] != string[right]:           
                break
            left -= 1
            right += 1
        return [left + 1, right]

File: 5-longest-palindromic-substring/5-longest-palindromic-substring.java
class Solution {
    public String longestPalindrome(String str) {
        int[] ansIdx = {0, 1};
        for (int i = 1; i < str.length(); i++) {
            int[] odd = getLongestPalindrom(str, i-1, i+1);
            int[] even = getLongestPalindrom(str, i-1, i);
            int[] longestIdx = (odd[1]-odd[0] > even[1]-even[0]) ? odd : even;
            ansIdx = (ansIdx[1]-ansIdx[0] > longestIdx[1]-longestIdx[0]) ?                          ansIdx : longestIdx;
        }
        return str.substring(ansIdx[0], ansIdx[1]);
    }
    
    public static int[] getLongestPalindrom(String str, int left, int right) {
        while (left >= 0 && right < str.length()) {
            if (str.charAt(left) != str.charAt(right)) {
                break;
            }
            left--;
            right++;
        }
        return new int[] {left + 1, right};
    }
}